[cssom] Proper serialization for border-radius and -webkit-border-radius

Loirooriol opened the pull request:
#### fd9594cdf418745718908fd1869a116307a46e7a
<pre>
[cssom] Proper serialization for border-radius and -webkit-border-radius
<a href="https://bugs.webkit.org/show_bug.cgi?id=183994">https://bugs.webkit.org/show_bug.cgi?id=183994</a>

Reviewed by Darin Adler.

The border-radius shorthand serialized incorrectly at specified-value
time. And -webkit-border-radius was never serialized.

The same serialization logic is used for both shorthands, except that if
border-radius would serialize as &apos;l1 l2&apos;, -webkit-border-radius chooses
&apos;l1 l2 l1&apos;, since &apos;-webkit-border-radius: l1 l2&apos; would be parsed like
&apos;border-radius: l1 / l2&apos;.

Tests: imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid.html
       imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid.html
       imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

* LayoutTests/fast/css/getComputedStyle/getComputedStyle-borderRadius-expected.txt:
Expect border radius longhands to always compute to a CSSPrimitiveValue.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid-expected.txt:
Update expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid.html:
Import changes from WPT.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid-expected.txt: Added.
Add expectations for new test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid.html: Added.
Add new test.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Update expectations.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::borderRadiusCornerValues):
(WebCore::borderRadiusCornerValue):
(WebCore::borderRadiusShorthandValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
Serialize -webkit-border-radius at computed-value time.
Change border radius longhands to use a Pair CSSPrimitiveValue.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::borderRadiusShorthandValue const):
Serialize -webkit-border-radius at specified-value time.
Fix serialization of border-radius at specified-value time.

* Source/WebCore/css/StyleProperties.h:
Add method.

Canonical link: <a href="https://commits.webkit.org/257041@main">https://commits.webkit.org/257041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/676467e8ceae484f2a00f29ec784a3235b6e3d5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97677 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107173 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167438 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101625 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7309 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35689 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90071 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103828 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103322 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5479 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84309 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32471 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87359 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89129 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75391 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/918 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20573 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/909 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22059 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4842 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5727 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44521 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41451 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -671,6 +671,11 @@ static Ref<CSSValueList> borderRadiusShorthandValue(const RenderStyle& style)
         list->append(WTFMove(verticalRadiiList));
+    else if (propertyID == CSSPropertyWebkitBorderRadius) {
+        auto* horizontals = downcast<CSSValueList>(list->item(0));
+        if (horizontals->length() == 2)
```
Can we base this check on the Boolean locals instead of reading back the length of the list?

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1242,6 +1243,67 @@ String StyleProperties::borderImagePropertyValue(const StylePropertyShorthand& s
+    for (unsigned i = 0; i < 4; ++i) {
+        auto value = getPropertyCSSValue(shorthand.properties()[i]);
+        ASSERT(!value || is<CSSPrimitiveValue>(*value));
+        if (!value || !is<CSSPrimitiveValue>(*value))
```
This can be written:
```
if (!is<CSSPrimitiveValue(value))
```
The `is` function builds in the null check.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1242,6 +1243,67 @@ String StyleProperties::borderImagePropertyValue(const StylePropertyShorthand& s
+            return String();
+
+        // FIXME: Remove this isCSSWideKeyword check after we do this consistently for all shorthands in getPropertyValue.
+        if (CSSValueID id = downcast<CSSPrimitiveValue>(*value).valueID(); isCSSWideKeyword(id)) {
```
`auto`?

Also since I prefer words to abbreviations I might name the local `keyword`.

darinadler commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -645,32 +642,42 @@ static Ref<CSSValueList> borderRadiusShorthandValue(const RenderStyle& style)
     bool showVerticalTopRight = showVerticalBottomRight || (style.borderTopRightRadius().height != style.borderTopLeftRadius().height);
 
     auto topLeftRadius = borderRadiusCornerValues(style.borderTopLeftRadius(), style);
+    CSSValue* topLeftRadiusX = topLeftRadius->first();
```
I would have used `auto` or `auto*`.

Loirooriol replied:
But then it becomes a `CSSPrimitiveValue*`, which somehow causes a compile error:

```
In file included from /app/webkit/Source/WebCore/css/CSSValuePool.h:28,
                 from /app/webkit/Source/WebCore/css/CSSValuePool.cpp:27,
                 from /app/webkit/WebKitBuild/Release/WebCore/DerivedSources/unified-sources/UnifiedSource-2f84417a-11.cpp:2:
/app/webkit/Source/WebCore/css/CSSPrimitiveValue.h:192:33: error: inline function ‘WebCore::CSSPrimitiveValue::operator T() const [with T = WTF::Ref<WebCore::CSSValue>]’ used but never defined [-Werror]
  192 |     template<typename T> inline operator T() const; // Defined in CSSPrimitiveValueMappings.h
      | 
```

Using `CSSValue*` it worked :shrug:

Loirooriol committed:
[cssom] Proper serialization for border-radius and -webkit-border-radius
https://bugs.webkit.org/show_bug.cgi?id=183994

Reviewed by Darin Adler.

The border-radius shorthand serialized incorrectly at specified-value
time. And -webkit-border-radius was never serialized.

The same serialization logic is used for both shorthands, except that if
border-radius would serialize as 'l1 l2', -webkit-border-radius chooses
'l1 l2 l1', since '-webkit-border-radius: l1 l2' would be parsed like
'border-radius: l1 / l2'.

Tests: imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid.html
       imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid.html
       imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks.html

* LayoutTests/fast/css/getComputedStyle/getComputedStyle-borderRadius-expected.txt:
Expect border radius longhands to always compute to a CSSPrimitiveValue.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid-expected.txt:
Update expectations.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/border-radius-valid.html:
Import changes from WPT.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid-expected.txt: Added.
Add expectations for new test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/parsing/webkit-border-radius-valid.html: Added.
Add new test.

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
Update expectations.

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::borderRadiusCornerValues):
(WebCore::borderRadiusCornerValue):
(WebCore::borderRadiusShorthandValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
Serialize -webkit-border-radius at computed-value time.
Change border radius longhands to use a Pair CSSPrimitiveValue.

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::borderRadiusShorthandValue const):
Serialize -webkit-border-radius at specified-value time.
Fix serialization of border-radius at specified-value time.

* Source/WebCore/css/StyleProperties.h:
Add method.

Canonical link: https://commits.webkit.org/257041@main
